AutomaticInterface/AutomaticInterface/RoslynExtensions.cs

```diff
@@ -71,6 +68,118 @@ public static List<string> GetUsings(this CompilationUnitSyntax root)
                 .Select(n => n.Name.ToString())
                 .ToList();
         }
-    }
-}
 
+        /// <summary>
+        /// Thanks to https://www.codeproject.com/Articles/871704/Roslyn-Code-Analysis-in-Easy-Samples-Part-2
+        /// </summary>
+        /// <param name="typeParameterSymbol"></param>
+        /// <returns></returns>
+        public static string GetWhereStatement(this ITypeParameterSymbol typeParameterSymbol)
+        {
+            string result = "where " + typeParameterSymbol.Name + " : ";
+
+            string constraints = "";
+
+            bool isFirstConstraint = true;
+
+            if (typeParameterSymbol.HasReferenceTypeConstraint)
+            {
+                constraints += "class";
+
+                isFirstConstraint = false;
+            }
+
+            if (typeParameterSymbol.HasValueTypeConstraint)
+            {
+                constraints += "struct";
+
+                isFirstConstraint = false;
+            }
+
+            foreach (var constraintType in typeParameterSymbol.ConstraintTypes)
+            {
+                // if not first constraint prepend with comma
+                if (!isFirstConstraint)
+                {
+                    constraints += ", ";
+                }
+                else
+                {
+                    isFirstConstraint = false;
+                }
+
+                constraints += constraintType.GetFullTypeString();
+            }
+
+            if (string.IsNullOrEmpty(constraints))
+                return null;
+
+            result += constraints;
+
+            return result;
+        }
+
+        public static string GetFullTypeString(this ITypeSymbol type)
+        {
+            string result =
+                type.Name +
+                type.GetTypeArgsStr(symbol => ((INamedTypeSymbol)symbol).TypeArguments);
+
+            return result;
+        }
+
+        static string GetTypeArgsStr
+        (
+            this ISymbol symbol,
+            Func<ISymbol, IEnumerable<ITypeSymbol>> typeArgGetter
+        )
+        {
+            IEnumerable<ITypeSymbol> typeArgs = typeArgGetter(symbol).ToList();
+
+            string result = "";
+
+            if (typeArgs.Any())
+            {
+                result += "<";
+
+                bool isFirstIteration = true;
+                foreach (ITypeSymbol typeArg in typeArgs)
+                {
+                    // insert comma if not first iteration                    
```

Comment:
maybe first create a list of strToAdd and then do a String.Join? Could be more readable

Reply:
Rewrited